### PR TITLE
Remove apm's node-gyp dependency

### LINKED ISSUE #18540: Mac: Script/bootstrap fails to install apm
[2018-12-04T18_33_48_069Z-debug.log](https://github.com/atom/atom/files/2645586/2018-12-04T18_33_48_069Z-debug.log)
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: https://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

When attempting to build atom on mac osx, I get the following error when it gets to the install apm step. This is when running `script/bootstrap` for the first time.
```
Error: The module '/Users/ryan/github/atom/atom/apm/node_modules/atom-package-manager/node_modules/git-utils/build/Release/git.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 64. This version of Node.js requires
NODE_MODULE_VERSION 57. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
```

### Steps to Reproduce

1. [First Step]
2. [Second Step]
3. [and so on...]

**Expected behavior:** Things to get bootstrapped

**Actual behavior:** Failed to install deps for apm install

**Reproduces how often:** 100%

### Versions

You can get this information from copy and pasting the output of `atom --version` and `apm --version` from the command line. Also, please include the OS and what version of the OS you're running.

Node:	v10.10.0
Npm:	v6.4.1

### Additional Information

Latest master

Any additional information, configuration or data that might be necessary to reproduce the issue.

### PR DESCRIPTION
**Pull request by winstliu**
None of the templates really apply here.

apm has long used a pinned version of node-gyp to install Node (or rather, Electron). The intent was always to update node-gyp every time npm was updated; however, this never happened and apm's version of node-gyp has become significantly dated as a result.

We now attempt to go back to how it was before and directly depend on npm's version of node-gyp.

For more information, see atom/apm#832.

Fixes #18540?

### CODE REVIEW

**Comment by 50Wliu** [comment]
`/bin/sh: /Users/vsts/agent/2.150.0/work/1/s/apm/node_modules/atom-package-manager/script/postinstall.sh: Permission denied`
Hmm...not sure how to fix this as all the permissions look correct on my end.

**Comment by 50Wliu** [comment]
:point_up: turns out you need to publish on a Unix machine in order for file permissions to stick around...the build's green now, though 😀

**Comment by vinkla** [comment]
I've tried running this locally on macOS and it seems to fix #18540

**Comment by rafeca** [comment]
Thanks!! 🚀

@50Wliu: Based on https://github.com/atom/apm/pull/832 comments, please update this PR once the stable release of `apm` is published

**Comment by 50Wliu** [comment]
🤦‍♂ forgot to publish from WSL again

**Comment by 50Wliu** [comment]
@rafeca I'm getting consistent timeouts on the x86 Windows build when running specs, and I'm not sure why. `script\test` doesn't even use apm.

**Comment by rafeca** [comment]
That's strange 🤷‍♂️ I'm gonna take a look at it, but I have no clue why would this happen 😟

**Comment by rafeca** [comment]
It seems like somewhere in the [`start.js`](https://github.com/atom/atom/blob/master/src/main-process/start.js) script the Atom executable gets halted, I'm adding some `console.log` statements to figure out where this is happening...

**Comment by rafeca** [comment]
@50Wliu looks like the process gets stalled between these lines: https://github.com/atom/atom/blob/bd05990ccf0ee6c08a9369083ad1c462f8725ce7/src/main-process/start.js#L1-L14

I'm adding more specific logging to see which is the module that's causing it

**Comment by rafeca** [comment]
After adding the debugging messages this is what gets logged in the CI now:

```
-- requiring module:  electron from D:\a\1\s\src\main-process\start.js
-- requiring module:  ../app from D:\a\1\s\out\Atom Dev\resources\electron.asar\browser\api\exports\electron.js
-- requiring module:  nslog from D:\a\1\s\src\main-process\start.js
-- requiring module:  ../build/Release/nslog.node from D:\a\1\s\node_modules\nslog\lib\nslog.js
-- requiring module:  electron from D:\a\1\s\out\Atom Dev\resources\electron.asar\browser\init.js
-- requiring module:  ../dialog from D:\a\1\s\out\Atom Dev\resources\electron.asar\browser\api\exports\electron.js
-- requiring module:  electron from D:\a\1\s\out\Atom Dev\resources\electron.asar\browser\api\dialog.js
-- requiring module:  ../app from D:\a\1\s\out\Atom Dev\resources\electron.asar\browser\api\exports\electron.js
-- requiring module:  ../browser-window from D:\a\1\s\out\Atom Dev\resources\electron.asar\browser\api\exports\electron.js
```

([source](https://github.visualstudio.com/Atom/_build/results?buildId=39605&view=logs&jobId=97a617bf-bcbd-5dfa-bba2-cfba2747b693&taskId=b3fc136d-32ef-515d-013e-5257567380ba&lineStart=25&lineEnd=26&colStart=1&colEnd=1)).

The [`browser-window.js`](https://github.com/electron/electron/blob/1-3-x/lib/browser/api/browser-window.js) file seems quite inoffensive, so this may not be telling us much about the root cause of the issue... Sorry if I haven't been able to provide much help here

**Comment by 50Wliu** [comment]
Oh, my hunch is that we're attempting to show a dialog box because of a failure, and then hanging after the dialog is successfully displayed.

On the Windows x64 build, here's what the require chain looks like:
windows-x64-requires

On the hanging Windows x86 build, here's what the require chain looks like:
windows-x86-requires

I expect apm was unable to correctly rebuild `nslog`'s dependencies.

**Comment by 50Wliu** [comment]
Bingo.
https://github.com/electron/electron/blob/2-1-x/lib/browser/init.js#L54-L57

**Comment by 50Wliu** [comment]
`Error: \\?\D:\a\1\s\node_modules\nslog\build\Release\nslog.node is not a valid Win32 application`

**Comment by rafeca** [comment]
Good find @50Wliu !

**Comment by 50Wliu** [comment]
@rafeca I think I'm going to try adding back the node-gyp dependency but updating it to 3.7.0, the latest version. Let me know if that sounds like a good plan to you.

**Comment by 50Wliu** [comment]
https://github.visualstudio.com/Atom/_build/results?buildId=40065&view=logs&jobId=97a617bf-bcbd-5dfa-bba2-cfba2747b693&taskId=ce76a027-d404-5bfb-7945-43613713e966&lineStart=246&lineEnd=246&colStart=1&colEnd=29
`npm verb cli   '--arch=x64',`
Well, just need to figure out why we're setting the arch incorrectly and we should be good.

**Comment by 50Wliu** [comment]
So.....
https://github.visualstudio.com/Atom/_build/results?buildId=40070&view=logs&jobId=97a617bf-bcbd-5dfa-bba2-cfba2747b693&taskId=ce76a027-d404-5bfb-7945-43613713e966&lineStart=22&lineEnd=23&colStart=1&colEnd=1
BUT
https://github.com/nodejs/node-gyp/blob/2b5ce52a26018692e89377b0504e7efa3b4900b6/bin/node-gyp.js#L43

Somehow the arch gets messed up when spawning the shell?

File: script/lib/run-apm-install.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ module.exports = function (packagePath, ci, stdioOptions) {
   installEnv.npm_config_target = CONFIG.appMetadata.electronVersion
   childProcess.execFileSync(
     CONFIG.getApmBinPath(),
-    ['--loglevel=error', ci ? 'ci' : 'install'],
+    [ci ? 'ci' : 'install'],
```
Comment (50Wliu) [inline script/lib/run-apm-install.js]: (apm doesn't recognize the --loglevel option)

**Comment by rafeca** [comment]
Wow! Awesome job @50Wliu ! ✨

How risky do you think is merging this? I'm asking because on Friday we're planning to roll the railcars for the next Atom release, so if there's a chance it'll break things (or native modules) I'd rather wait until Monday to merge it to minimize the potential release blockers.

**Comment by rafeca** [comment]
LOL! why does my comment appear above yours? 🤯

**Comment by 50Wliu** [comment]
Yeah, we can totally wait until Monday to be safe.

**Pull request merged**